### ISSUE TITLE
[core] Avoid parsing rulesets multiple times

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#724]
We currently parse XMLs in a greedy and not very efficient way.

1. Upon finding a ruleset reference (ended in .xml), we go and parse that file completely before continuing.
2. Upon finding a rule reference from another ruleset (ended in .xml/rule-name) we go and parse that file completely, ignoring everything but the one rule we were looking for.

This means that, a ruleset such as:

https://github.com/pmd/pmd/blob/b159143a06cd5c6ec3c6263a9518938ea4f223d2/pmd-java/src/main/resources/rulesets/java/coupling.xml#L1-L20

Will parse completely once looking for `category/java/bestpractices.xml`, and then will proceed to completely parse `category/java/design.xml` **4 times**, to get the other rules.

Having multiple such references is common, and will be far more with the ruleset reorganization, since categories are fewer, larger, and intended to be broken up and referenced rule by rule from rulesets.

This also means, it's possible to produce cycles in the ruleset references and produce infinite loops (which end in Stackoverflows / OutOfMemory errors).

I believe we should rework the rule parsing in order to:

1. Parse each file exactly once
2. Avoid parsing beyond what's needed
3. Detect and report properly cycles

One possible approach to this would be:

1. Start with an empty set of open and a sorted list of closed `RuleSetReferenceId` containing all rulesets indicated by user input, with the "include all" flag set (more on sorting later on)
2. Remove the first `RuleSetReferenceId` from closed, put it into open, and get a parser for it
    1. If the "include all" flag is set, parse the file completely. Rule definitions are loaded, rule / ruleset references, are simply translated into `RuleSetReferenceId` and if present in open, an exception is thrown (cycle!) otherwise check if closed has another reference to this xml, and merge it or add it directly (more on merging below)
    2. If the "include all" flag is not set, parse until we find any required rule name and mark it as loaded from `RuleSetReferenceId`. Rule definitions are loaded, rule / ruleset references, are simply translated into `RuleSetReferenceId` and if present in open, an exception is thrown (cycle!) otherwise check if closed has another reference to this xml, and merge it or add it directly (more on merging below)
        1. If not all references from `RuleSetReferenceId` are loaded, go back to step 2.ii, otherwise just stop parsing.
3. Go back to step 2 until open is empty

Notice that the 2 scenarios (include all, described in 2.i and not, described in 2.ii) can be unified into a single one if the `haveAllRequiredRulesBeenLoaded()` check for `RuleSetReferenceId` returns always false in the `include all` scenario.

On sorting:
- we should prioritize ruleset references with the "include all" flag. That way we are able to include *ALL* references to rules within a single ruleset xml *before* parsing it.
- For non "include all" references, order is probably irrelevant, but prioritizing those with more rules to be loaded could be beneficial under complex ruleset graphs

On merging:
 - `RuleSetReferenceId` should be reworked to be able to reference, 1, many or all rules within a ruleset; and if not including all, keep track of which ones have been loaded. This way we can merge 2 references to different rules inside the same xml to be parsed together.

The one scenario that wouldn't be supported, would be a rule within a ruleset that references another rule within the same ruleset. This would be detected as a cycle, which is technically correct since it's a loop in the ruleset graph. We *could* support it if it was later in the file, but such distinction seems broken to me, I'd rather just throw for cycle.

This refactor would pose a good chance to split rule parsing from ruleset generation, as it has proved to be troublesome as show on https://github.com/pmd/pmd/pull/731/files#r152080633

**Comment by oowekyala** [comment]
@jsotuyod are you working on this?

About your algorithm, what if several different rulesets reference the same rule with different (or same) properties? I think your algo would report a cycle, even though it probably shouldn't be an error.

Also when you say 

> "parse until we find any required rule name ... otherwise just stop parsing."

Do you mean to use a SAX API?


I'm not very comfortable with adding so much power to RulesetReferenceId. Ideally for me, it should remain immutable, which would make it much simpler to follow. I'd like more a wrapper class, private to RulesetFactory, because the loaded rule tracking is only relevant to that algorithm.

**Comment by jsotuyod** [comment]
@oowekyala not at the moment. I want to first finish:

1. reviewing open PRs (specialy a couple of yours that are waiting for some time now)
2. finish my work on #904 so we can merge it into 6.2.0

> About your algorithm, what if several different rulesets reference the same rule with different (or same) properties? I think your algo would report a cycle, even though it probably shouldn't be an error.

We don't currently properly support this either. We assume there is a single rule with a given name at all times.
https://github.com/pmd/pmd/blob/6943803ccb8d2bf8604b26b7865e270f18f0d87d/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java#L448-L454

Moreover, having a single rule with different configurations at the same time seems weird. I don't think we have a proper use case for this.

> Do you mean to use a SAX API?

Yes, that's exactly what I was implying.

Using SAX would avoid creation of node elements we don't really need, avoid reading of files up to the end, and be generally more efficient. This of course can be done on a separate step of the refactoring, as most of the gains will come from simply not parsing files over and over.

> I'm not very comfortable with adding so much power to RulesetReferenceId. Ideally for me, it should remain immutable, which would make it much simpler to follow. I'd like more a wrapper class, private to RulesetFactory, because the loaded rule tracking is only relevant to that algorithm.

Maybe. It's not so much logic to be added, but it could be an option.

**Comment by jsotuyod** [comment]
Status update: I'm actively working on this.

- I decided on StAX API to be able to stop parsing prematurely.
- I decided not to change most pre-existing classes, but rather deprecate them in favor of new ones (already marked as internal in most cases as per #995).
- I started with the Cursor API to keep it lightweight, however I came across some difficulties (ie: setting rule properties needs to first load the whole rule) that I'm still considering. I *may* change to an Iterator API.

**Comment by oowekyala** [comment]
@jsotuyod If we change the schema in 7.0.0, the RulesetFactory/ RuleFactory will need a rewrite, which means it will be a good time to implement this. 

I'm still not sure whether a SAX parsing really adds more value compared to the complexity it adds. For example, a DOM-based approach like the following may prove sufficient while being very simple to implement:

Suppose we have the following wrapper objects:

```java
class RulesetDomWrapper {
  Document rulesetDom;
  Map<String, RuleDomWrapper> ruleNameToRuleSupplier;

 RulesetDomWrapper(Document rsetXml) {
   // iterate the "rule" elements and populate the map with the wrappers
 }

 Rule newRuleByName(String name) {
    return ruleNameToRuleSupplier.get(name).toRule();
  }


}
/** Wraps a rule element and can convert it to a Rule if needed. The conversion is done lazily.*/
class RuleDomWrapper {
   Element ruleElement;
  
  RuleDomWrapper(Element ruleElt) {
    ruleElement = ruleElt;
  }

   /** Builds a new rule based on the DOM element. */
   Rule toRule() {
     // converts the element to a rule (using RuleFactory)
     // the actual DOM exploration only needs to be done once,
     // after that, we can cache a fresh rule and just deep copy it
     // on subsequent calls.
   }
}
```
Basically these wrap the DOM and only convert Elements to Rule if needed.
When building a ruleset, each time we encounter a rule reference to an external ruleset, we read the ruleset file and wrap the dom in a RulesetDomWrapper. Then we get the rule by name. The ruleset's wrapper is cached and reused the next time we need a rule from the same ruleset.

This approach would:
* avoid parsing rulesets multiple times
* avoid parsing rules we don't need
* have minimal impact on our existing codebase compared to a SAX rewrite

Of course SAX allows us to stop XML parsing once we found all the rules we need, but realistically, I would guess the performance bottleneck is *not* the actual XML parsing (since ruleset files are not huge anyway), but the fact that for now, we parse and throw away the whole ruleset for each rule. I can't see other real advantages to using SAX, but maybe you have more?

Of course this approach would need to be fleshed out to detect circular references and deal with reference indirection, but I don't think that would be too hard to implement.

**Comment by jsotuyod** [comment]
@oowekyala I already have most of this on my local branch already in place. I'll try to get a working version up soon. The difference is actually quite significant, but I acknowledge I haven't split the impact between StAX vs just parsing once each file....

I'll try to break it down into stages and validate the premise.
